### PR TITLE
Update to Webpack 3. Almost all dependency are updated.

### PR DESCRIPTION
**Pull request by lionel-bijaoui**
Reorganisation of the config files. Now use Prettier, so most files are will need to be updated to validate. This is stil an early WIP.

Known bugs: validation is broken since rewritten in without "module.exports"

### CODE REVIEW

**Comment by lionel-bijaoui** [comment]
* Many things are still to fix and change. My goal is passing eslint with prettier rules on all files in `src` and `dev` while having everything work as before.
* Only `npm run dev` was tested. `npm run build` still need to be updated.
* I have broken the validation. We can't use `module.exports` anymore, and while changing it I clearly broke it. @icebob I will need you help to fix it.
* I have moved files around (it copy the [vue + webpack template](https://github.com/vuejs-templates/webpack)). I tried some refactoring on the dev files, putting things in common and unifying as much as I was able to.
* Constructive feedback and new ideas are welcome.

**Comment by zoul0813** [comment]
Will review tomorrow morning

**Comment by zoul0813** [comment]
I'm not sure if I'm comfortable with merging this in, as it's currently in an unusable state.  The broken validators are a major issue, for example, as is the broken `npm run build`.

I like the direction you're heading in though.

Perhaps we should have two branches for v3? One as a 'v3-master' and another as a 'v3-dev'.  This way we can merge "broken" code in so everyone can contribute to fixing it, and another for "stable" v3 changes?

**Comment by lionel-bijaoui** [comment]
This is not supposed to be ready to merge yet.
This is a (bleeding edge) proof of concept and I will need more time to finish it then I will ask for a merge.
As i said :
> My goal is passing eslint with prettier rules on all files in src and dev while having everything work as before.

I want feedback from the collaborator on the direction I going with this.
I don't want to work on my side and then realized no one like the way I organized the files, the prettier config or something like that.

My biggest problem is my lack of time, since at work, we are very close to the big launch + Holidays coming.
I'm doing my best to create a solid base for the future of VFG, please give me time ⌛️ and feedback ⚡️

**Comment by lionel-bijaoui** [comment]
I'm back ! I have just updated everything and I still need to update this branch with the latest correction from v2.
The good news is that I will have time in two weeks to work on this almost full time.

**Comment by lionel-bijaoui** [comment]
That's done, the v3 is up to date with the latest fix from v2.
I'm going to fix the build process now.

**Comment by lionel-bijaoui** [comment]
Little update:
- it use jsdom instead of phantomJs, so it is way faster to initialize,
- I have skipped most of the test, only enabling one at the time and fixing all errors,
- so far I only have the pure JS test working
```
  65 passing (180ms)
  385 pending
```
- the hard part is coming, I have to use test-utils to wrap the Vue components and adapt the test with this new tools. I will first try to fix the test without it so I don't spend too much time on it.

**Comment by lionel-bijaoui** [comment]
# Progress Update
```
  154 passing (534ms)
  295 pending
```
I had to skip some test (6 for now).
- index.spec.js - no skip
- validators.spec.js - no skip
- schema.spec.js - no skip
- VueFormGenerator.spec.js - 4 skip
- abstractField.spec.js - 1 skip
- fieldCheckbox.spec.js - 1 skip

**Comment by lionel-bijaoui** [comment]
# Progress Update (19/02/2018)
```
  154 passing (534ms)
  295 pending
```
I had to skip some test (6 for now).
- index.spec.js - no skip
- validators.spec.js - no skip
- schema.spec.js - no skip
- VueFormGenerator.spec.js - 4 skip
- abstractField.spec.js - 1 skip
- fieldCheckbox.spec.js - 1 skip

**Comment by lionel-bijaoui** [comment]
# Progress Update
```
  257 passing (2s)
  193 pending
```
I'm hitting the first [real problem](https://github.com/rstacruz/jsdom-global/issues/15#issuecomment-368816528) with jQuery based fields. I'm going to skip them for the moment.
- index.spec.js - no skip
- validators.spec.js - no skip
- schema.spec.js - no skip
- VueFormGenerator.spec.js - 4 skip
- abstractField.spec.js - 1 skip
- fieldCheckbox.spec.js - 1 skip
- fieldChecklist.spec.js -2 skip
- fieldCleave.spec.js - no skip
- fieldDateTimePicker.spec.js - full skip (jQuery based)
- fieldGoogleAddress.spec.js - no skip (will need some work later to make it [work correctly](https://stackoverflow.com/questions/41958271/load-google-maps-api-in-jsdom))

**Comment by zoul0813** [comment]
I feel like anything relying on jQuery should be reworked to not need jQuery, reducing the deps for VFG... as not everyone uses jQuery. I think we should also rework the optional fields to be separate packages to allow devs to pick and chose which ones they want to reduce the overall size of VFG.  

Let me know if there’s anything I can do to help.

**Comment by lionel-bijaoui** [comment]
@zoul0813 There is already vanilla js alternative and jQuery is optional. But I absolutely agree, all fields should be their own project.
It would greatly reduce the weight on our shoulder and make thing clearer (separated issues and all). @icebob was against this idea so it doesn't seem to be something that is going to happen.

I'm doing my best to finish the test as far as I can go and merge. Then I think every collaborator will be able to help. Still a looooot of work toward this new version, this is exciting and frightening 😀 

Thank you !

**Comment by ChadTaljaardt** [comment]
Should this not be updated to webpack 4 instead of 3?

**Comment by lionel-bijaoui** [comment]
@ChadTaljaardt This project started before the v4 was even announced. I feel like it will require a lot of changes and I'm not sure the plugin ecosystem is up to date already.
Maybe later this year will we be able to update. I don't want to lose more time on this.

**Comment by zoul0813** [comment]
I thought @icebob said splitting the optional fields out was a good idea - could be mistaken, lots of discussions going on.  I was under the impression the new group organization was setup to help facilitate the split of the optional fields into different packages (as well as create more projects with similar goals).

**Comment by lionel-bijaoui** [comment]
I may have misunderstood. You are right, this new organization is perfect for that. Maybe @icebob changed his mind ? I sure hope so !

**Comment by lionel-bijaoui** [comment]
# Progress Update
```
  362 passing (3s)
  89 pending
```
- index.spec.js - no skip
- validators.spec.js - no skip
- schema.spec.js - no skip
- VueFormGenerator.spec.js - 4 skip
- abstractField.spec.js - 1 skip
- fieldCheckbox.spec.js - 1 skip
- fieldChecklist.spec.js -2 skip
- fieldCleave.spec.js - no skip
- fieldDateTimePicker.spec.js - 1 skip (jQuery based)
- fieldGoogleAddress.spec.js - no skip (will need some work later to make it [work correctly](https://stackoverflow.com/questions/41958271/load-google-maps-api-in-jsdom))
- fieldImage.spec.js - 1 skip
- fieldInput.spec.js - no skip
- fieldLabel.spec.js - no skip
- fieldMasked.spec.js - 1 skip (jQuery based)
- fieldNoUiSlider.spec.js - 1 skip
- fieldPikaday.spec.js - 1 skip
- fieldRadios.spec.js - no skip
- fieldRangeSlider.spec.js - 3 skip (jQuery based)
- fieldSelect.spec.js - no skip

**Comment by zoul0813** [comment]
@lionel-bijaoui well, if he hasn't ... maybe we can do a "best of both worlds" solution, and split the optional fields up into separate packages anyway ... then have two versions of VFG ... a VFG Lite and a VFG Full, where VFG Full just includes all the optional packages in it's package.json ... this would then allow us to have the stream lined separate packages for easier development and testing of advanced fields, as well as provide devs with a "lite" version and a "full" version ... you could then start with the "lite" version and add any extra optional fields you wish, or just use the "full" version and get everything with a single package dep.

The "lite" version would just include basic form fields, anything that's part of the HTML5 standard (input, select, check, radio, buttons, hidden fields, etc).

@icebob thoughts?

**Comment by lionel-bijaoui** [comment]
@zoul0813 It is already like that, you've got the core or full version. But from experience, we need an "engine" version without any field at all.

**Comment by zoul0813** [comment]
@lionel-bijaoui the difference in what I'm proposing is that if you enable "full" you get it all ... you can't selectively decide that you want to use the "core" and then add in 1-2 optional fields ...

Splitting the optional fields up into there own packages would allow me to just `npm install --save-dev vfg-core vfg-optional-field-name`, and then I'd have the core and only that single optional field, as opposed to all of the optional fields.

**Comment by icebob** [comment]
Hi guys,

sorry for late. Yes, I support to split optional modules to separated packages in v3. And as @zoul0813 said, we can create a "full" bundle, which install all optional modules as npm dependencies.

**Comment by lionel-bijaoui** [comment]
@icebob I'm so happy to hear that ! Will it use git submodule or separate package ?

**Comment by icebob** [comment]
I think separate npm packages in separate github repos under vue-generators org.

**Comment by icebob** [comment]
I will try to set npm publishing [on travis](https://docs.travis-ci.com/user/deployment/npm/), so every team member will able to release new versions without me. _(Nowadays I'm very busy, sorry)_

**Comment by lionel-bijaoui** [comment]
# Progress Update
```
  416 passing (4s)
  35 pending
```
I worked on the overhaul of code coverage. I'm hopping this will pass Travis. Only one file to do ! I'm seeing the light at the end of the tunnel *u*
- index.spec.js - no skip
- validators.spec.js - no skip
- schema.spec.js - no skip
- VueFormGenerator.spec.js - 4 skip
- abstractField.spec.js - 1 skip
- fieldCheckbox.spec.js - 1 skip
- fieldChecklist.spec.js -2 skip
- fieldCleave.spec.js - no skip
- fieldDateTimePicker.spec.js - 1 skip (jQuery based)
- fieldGoogleAddress.spec.js - no skip (will need some work later to make it [work correctly](https://stackoverflow.com/questions/41958271/load-google-maps-api-in-jsdom))
- fieldImage.spec.js - 1 skip
- fieldInput.spec.js - no skip
- fieldLabel.spec.js - no skip
- fieldMasked.spec.js - 1 skip (jQuery based)
- fieldNoUiSlider.spec.js - 1 skip
- fieldPikaday.spec.js - 1 skip
- fieldRadios.spec.js - no skip
- fieldRangeSlider.spec.js - 3 skip (jQuery based)
- fieldSelect.spec.js - no skip
- fieldSelectEx.spec.js - 4 skip (jQuery based)
- fieldSpectrum.spec.js - 3 skip (jQuery based)
- fieldStaticMap.spec.js - no skip
- fieldSubmit.spec.js - 3 skip
- fieldSwitch.spec.js - no skip
- fieldUpload.spec.js - no skip

**Comment by lionel-bijaoui** [comment]
# Progress Update
```
  425 passing (4s)
  26 pending
```
I cleaned the files and remove obsolete function. I'm not going to do any more work on the test. My goal now is to make the build successful.
- index.spec.js - no skip
- validators.spec.js - no skip
- schema.spec.js - no skip
- VueFormGenerator.spec.js - 4 skip
- abstractField.spec.js - 1 skip
- fieldCheckbox.spec.js - 1 skip
- fieldChecklist.spec.js -2 skip
- fieldCleave.spec.js - no skip
- fieldDateTimePicker.spec.js - 1 skip (jQuery based)
- fieldGoogleAddress.spec.js - no skip (will need some work later to make it [work correctly](https://stackoverflow.com/questions/41958271/load-google-maps-api-in-jsdom))
- fieldImage.spec.js - 1 skip
- fieldInput.spec.js - no skip
- fieldLabel.spec.js - no skip
- fieldMasked.spec.js - 1 skip (jQuery based)
- fieldNoUiSlider.spec.js - 1 skip
- fieldPikaday.spec.js - 1 skip
- fieldRadios.spec.js - no skip
- fieldRangeSlider.spec.js - 3 skip (jQuery based)
- fieldSelect.spec.js - no skip
- fieldSelectEx.spec.js - 4 skip (jQuery based)
- fieldSpectrum.spec.js - 3 skip (jQuery based)
- fieldStaticMap.spec.js - no skip
- fieldSubmit.spec.js - 3 skip
- fieldSwitch.spec.js - no skip
- fieldUpload.spec.js - no skip
-fieldVueMultiSelect.spec.js - no skip

**Comment by lionel-bijaoui** [comment]
@icebob I don't know how to activate Codacy for this branch, but expect for that, everything is ready for the merge on my side.
Dev, build, test are working, travis is happy and coveralls is working.

**Comment by icebob** [comment]
@lionel-bijaoui Great job! Thank you!

File: config/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// see http://vuejs-templates.github.io/webpack for documentation.
+let path = require("path");
+
+module.exports = {
+	build: {
+		env: { NODE_ENV: "\"production\"" },
+		index: path.resolve(__dirname, "../../api/src/Template/Layout/layout_a.ctp"),
+		assetsRoot: path.resolve(__dirname, "../../api/webroot/"),
```
Comment (icebob) [inline config/index.js]: What are these paths?

Comment (lionel-bijaoui) [reply config/index.js]: oups sorry ! I based my work on an existing project. I will fix that ASAP

**Pull request merged**